mlflow/pyspark/ml/__init__.py

```diff
@@ -132,6 +136,27 @@ def _get_instance_param_map(instance):
     return expanded_param_map
 
 
+def _get_pipeline_stage_hierarchy_and_params(pipeline):
+    from pyspark.ml import Pipeline
+    """
+    :param pipeline_stages: stages of a pyspark ML pipeline
+    :return: return a tuple of (stage_hierarchy, params_map)
+    """
+    stage_hierarchy = []
+    param_map = {}
+    pipeline_stages = pipeline.getStages()
+    for stage in pipeline_stages:
+        if isinstance(stage, Pipeline):
+            hierarchy_elem, stage_param_map = _get_pipeline_stage_hierarchy_and_params(stage)
+        else:
+            hierarchy_elem = stage.uid
+            stage_param_map = {f'{stage.uid}.{k}': v
+                               for k, v in _get_instance_param_map(stage).items()}
+        stage_hierarchy.append(hierarchy_elem)
+        param_map.update(stage_param_map)
+    return {pipeline.uid: stage_hierarchy}, param_map
```

Comment:
@WeichenXu123 One thing I noticed about this construction is that we log the parameters for each nested pipeline stage, but the `stages` parameter for a nested pipeline only refers to the top-level pipeline (e.g. `[Tokenizer_112334721b1b, Pipeline_07c6a754c9a8]`).

This makes it a bit difficult to figure out where the stages of `Pipeline_07c6a754c9a8`, such as `HashingTF_59b3cf717aab` and `LogisticRegression_3b5baf5ff098` in the example you linked offline, come from.

Could we consider updating the `stages` parameter so that it expands each pipeline?

E.g. `{"stages": [Tokenizer_112334721b1b, Pipeline_07c6a754c9a8]}` would become `{"stages": [Tokenizer_112334721b1b, [HashingTF_59b3cf717aab, LogisticRegression_3b5baf5ff098]}`

Reply:
If the pipeline uid is important, we could also include it in the `stages` hierarchy  (e.g. `Pipeline_07c6a754c9a8: [HashingTF_59b3cf717aab, LogisticRegression_3b5baf5ff098]`

Reply:
Then it become the same with `pipeline_hyerarchy.json`, so shall we skip log `stages` param ?

Reply:
If anything, I think we could omit the hierarchy artifact and keep the stage parameter. Let me know if you disagree!

Reply:
Should we keep the hierarchy artifact, the stages hierarchy tends to be long and easily exceeds the maximum param value length (250) and will be truncated?